### PR TITLE
PMM-8770 enable safe metrics collection

### PR DESCRIPTION
**Pull request by percona-csalguero**
[PMM-8770: Enable safe Collection metrics collection for MongoDB](https://jira.percona.com/browse/PMM-8770) 

Feature Build: [PR #2154](https://github.com/Percona-Lab/pmm-submodules/pull/2154)

### CODE REVIEW

File: exporter/indexstats_collector.go [hunk 1]
```diff
@@ -69,27 +70,29 @@ func (d *indexstatsCollector) Collect(ch chan<- prometheus.Metric) {
 		cursor, err := d.client.Database(database).Collection(collection).Aggregate(d.ctx, mongo.Pipeline{aggregation})
 		if err != nil {
 			d.logger.Errorf("cannot get $indexStats cursor for collection %s.%s: %s", database, collection, err)
+
```
Comment (ShashankSinha252) [inline exporter/indexstats_collector.go:73]: Unnecessary whitespace

File: exporter/indexstats_collector.go [hunk 2]
```diff
@@ -69,27 +70,29 @@ func (d *indexstatsCollector) Collect(ch chan<- prometheus.Metric) {
...
 			continue
 		}
 
 		var stats []bson.M
 		if err = cursor.All(d.ctx, &stats); err != nil {
 			d.logger.Errorf("cannot get $indexStats for collection %s.%s: %s", database, collection, err)
+
```
Comment (ShashankSinha252) [inline exporter/indexstats_collector.go:80]: Unnecessary whitespace

File: exporter/collstats_collector.go [hunk 3]
```diff
@@ -82,12 +83,14 @@ func (d *collstatsCollector) Collect(ch chan<- prometheus.Metric) {
 		cursor, err := d.client.Database(database).Collection(collection).Aggregate(d.ctx, mongo.Pipeline{aggregation, project})
 		if err != nil {
 			d.logger.Errorf("cannot get $collstats cursor for collection %s.%s: %s", database, collection, err)
+
```
Comment (ShashankSinha252) [inline exporter/collstats_collector.go:86]: Unnecessary whitespace

File: exporter/collstats_collector.go [hunk 4]
```diff
@@ -82,12 +83,14 @@ func (d *collstatsCollector) Collect(ch chan<- prometheus.Metric) {
...
 			continue
 		}
 
 		var stats []bson.M
 		if err = cursor.All(d.ctx, &stats); err != nil {
 			d.logger.Errorf("cannot get $collstats for collection %s.%s: %s", database, collection, err)
+
```
Comment (ShashankSinha252) [inline exporter/collstats_collector.go:93]: Unnecessary whitespace

File: exporter/indexstats_collector.go [hunk 5]
```diff
@@ -41,26 +41,27 @@ func (d *indexstatsCollector) Describe(ch chan<- *prometheus.Desc) {
...
 	if d.discoveringMode {
-		databases := map[string][]string{}
-		for _, dbCollection := range d.collections {
-			parts := strings.Split(dbCollection, ".")
-			if _, ok := databases[parts[0]]; !ok {
-				db := parts[0]
-				databases[db], _ = d.client.Database(parts[0]).ListCollectionNames(d.ctx, bson.D{})
-			}
+		namespaces, err := listAllCollections(d.ctx, d.client, d.collections)
```
Comment (ShashankSinha252) [inline exporter/indexstats_collector.go:47]: d.collections is used as regex pattern. Is this expectation correct from input perspective?

File: main.go [hunk 6]
```diff
@@ -46,6 +46,7 @@ type GlobalFlags struct {
 	DisableDiagnosticData   bool   `name:"no-collector.diagnosticdata" help:"Disable collecting metrics from getDiagnosticData"`
 	DisableReplicasetStatus bool   `name:"no-collector.replicasetstatus" help:"Disable collecting metrics from replSetGetStatus"`
 	EnableDBStats           bool   `name:"collector.dbstats" help:"Enable collecting metrics from dbStats"`
+	CollStatsLimit          int    `name:"collector.collstats-limit" help:"Enable collstats and indexstats collector only if there are less than <n> collections. -1=No limit" default:"100"`
```
Comment (BupycHuk) [inline main.go:49]: I would set no limit by default.

Comment (BupycHuk) [reply main.go:49]: And if we need limit in PMM set it there.

**Review by BupycHuk — APPROVED** [review]

**Comment by BupycHuk** [comment]
Please fix all golint issues

File: exporter/exporter.go [hunk 7]
```diff
@@ -78,21 +82,33 @@ func New(opts *Opts) *Exporter {
...
+		totalCollectionsCount: -1, // not calculated yet. waiting the db connection.
 	}
 	// Try initial connect. Connection will be retried with every scrape.
 	go func() {
-		if _, err := exp.getClient(ctx); err != nil {
+		_, err := exp.getClient(ctx)
+		if err != nil {
 			exp.logger.Errorf("Cannot connect to MongoDB: %v", err)
+
+			return
```
[the remark this answers is not in the document]
Comment (percona-csalguero) [reply exporter/exporter.go:98]: not really

File: exporter/collstats_collector.go (hunk 3 above)
Comment (percona-csalguero) [reply exporter/collstats_collector.go:86]: For the linter

File: exporter/indexstats_collector.go (hunk 5 above)
Comment (percona-csalguero) [reply exporter/indexstats_collector.go:47]: Yes, it is a filter

File: exporter/exporter.go [hunk 8]
```diff
@@ -229,6 +256,16 @@ func (e *Exporter) Handler() http.Handler {
 
 			return
 		}
+
+		if e.getTotalCollectionsCount() < 0 {
+			count, err := allCollectionsCount(ctx, client, nil)
+			if err == nil {
+				e.lock.Lock()
```
[the remark this answers is not in the document]
Comment (percona-csalguero) [reply exporter/exporter.go:261]: It is used only once

**Review by JiriCtvrtka — APPROVED** [review]

**Review by ShashankSinha252 — APPROVED** [review]
